### ISSUE TITLE
Enhancement: Allow 3rd party plugin support without that plugin needing webonyx/graphql-php

### ISSUE THREAD

**Issue by Tam** [markhuot/craftql#33]
It would be great if 3rd party plugins didn't need `webonyx/graphql-php` as a required dependency if they want to add CraftQL support. I think a lot more plugins would be up for supporting CraftQL if doing so didn't add any extra bloat (and some docs on how to add support wouldn't go amiss 😄).

I was thinking something along the lines of (example from [SimpleMap](https://github.com/ethercreative/simplemap/blob/v3/src/fields/MapField.php#L563)):

With `graphql-php`:
```php
return new ObjectType([
	'name' => 'SimpleMap',
	'fields' => [
		'lat' => Type::string(),
		'lng' => Type::string(),
		'zoom' => Type::string(),
		'address' => Type::string(),
		'parts' => [
			'type' => new ObjectType([
				'name' => 'SimpleMapParts',
				'fields' => $parts,
			]),
		],
	],
]);
```

My proposal sans `graphql-php`:
```php
return [
	'ObjectType',
	'name' => 'SimpleMap',
	'fields' => [
		'lat' => 'String',
		'lng' => 'String',
		'zoom' => 'String',
		'address' => 'String',
		'parts' => [
			'type' => [
				'ObjectType',
				'name' => 'SimpleMapParts',
				'fields' => $parts,
			],
		],
	],
];
```

The type simply becomes a string. For types that require additional parameters, the type would be the first, keyless, value of an array.

I'm sure there are more complex query field setups that this wouldn't work for, but I think it's a good jumping off point!

**Comment by markhuot** [comment]
Thanks for the feedback @Tam. I _completely_ agree that consuming plugins shouldn't have to depend on `webonyx/graphql-php`. Depending on the specific implementation that CraftQL chose makes it much more brittle.

That said, I don't know that `string` types is how I would go about this. Instead, it may make sense for CraftQL to expose its own types (that are just proxies for `webonyx/graphql-php` types, today).

```php
return new \markhuot\CraftQL\Types\ObjectType([
	'name' => 'SimpleMap',
	'fields' => [
		'lat' => \markhuot\CraftQL\Type::string(),
		'lng' => \markhuot\CraftQL\Type::string(),
		'zoom' => \markhuot\CraftQL\Type::string(),
		'address' => \markhuot\CraftQL\Type::string(),
		'parts' => [
			'type' => new \markhuot\CraftQL\Types\ObjectType([
				'name' => 'SimpleMapParts',
				'fields' => $parts,
			]),
		],
	],
]);
```

I understand, this still requires the SimpleMap plugin to depend on an external dependency for static analysis, though, so it's not 100% ideal. Let me poke around and see what others are doing in similar situations.

---

Crazy thought, I've entertained the idea of abstracting it _all_ away into some sort of builder DSL… is this any better?

```php
$parts = new \markhuot\CraftQL\Object('parts');

foreach (MapService::$parts as $part) {
    $parts->addStringField($part);
    $parts->addStringField($part . '_short');
}

return (new \markhuot\CraftQL\Object('SimpleMap'))
    ->addStringField('lat')
    ->addStringField('lng')
    ->addStringField('zoom')
    ->addStringField('address')
    ->addObjectField($parts)
]);
```

**Comment by markhuot** [comment]
Oh, as for documentation, yup! It'll come with a 1.0 release some time early next year. You're quite the early adopter so _thanks_ for talking this through with me. I'll be sure to let you know once I have some docs written.

**Comment by Tam** [comment]
I do like the DSL idea!

Perhaps having a dedicated file for the integration that would only be loaded by CraftQL, rather than a function on the field. That way, if the necessary CraftQL classes don't exist it shouldn't matter because that file isn't loaded? (I think that would work, but don't hold me to it!)

**Comment by markhuot** [comment]
That makes sense @Tam, thank you for thinking this through with me. I'll get back to you on a better approach here.

**Comment by markhuot** [comment]
What about `Event::on` instead? Maybe adding the following to your plugin `init`,

```php
Event::on(
    ether\simplemap\fields\MapField::class,
    \markhuot\CraftQL\CraftQL::EVENT_GET_FIELD_SCHEMA,
    [ether\simplemap\fields\GraphQLMapping::class, 'onGetFieldSchema']
)
```

Then, inside `GraphQLMapping::onGetFieldSchema` you could,

```php
class GraphQLMapping {
    static function onGetFieldSchema(/* Optional: \markhuot\CraftQL\Events\GetFieldSchema */ $event) {
        $parts = $event->builder->newObjectType('SimpleMapParts');

        foreach (MapService::$parts as $part) {
            $parts->addStringField($part);
            $parts->addStringField($part . '_short');
        }

        return $event->builder->newObjectType('SimpleMap')
            ->addStringField('lat')
            ->addStringField('lng')
            ->addStringField('zoom')
            ->addStringField('address')
            ->addObjectField('parts', $parts)
        ]);
    }
}
```

Or, you could do this all in a closure within the `init` too, the placement would really be up to the consumer.

**Comment by Tam** [comment]
Doing it event based would make it *very* Craft 3! Unfortunately, that would mean having CraftQL as a dependency to be able to hook into the event.

I really like the idea of simply passing the schema builder as an argument. Modifying the existing `getGraphQLQueryFields` method to accept the schema builder as an argument would make integration dependency free!

**Comment by markhuot** [comment]
Hrm, the only reference to `markhuot\CraftQL`, right now, is `CraftQL::EVENT_GET_FIELD_SCHEMA`, which resolves to a string. We _could_ specify the string directly. Something like this,…

```php
Event::on(
    ether\simplemap\fields\MapField::class,
    'craftQlGetFieldSchema',
    [ether\simplemap\fields\GraphQLMapping::class, 'onGetFieldSchema']
)
```

This is sort of how Yii recommends doing it too, http://www.yiiframework.com/doc-2.0/guide-concept-events.html#global-events

**Comment by markhuot** [comment]
This is still on a feature branch because I'm going to "dog food" it and update all my internal fields to the new builder approach. But… you can see what I'm thinking with the updated docs, here:

https://github.com/markhuot/craftql/tree/event-registration#third-pary-field-support

**Comment by Tam** [comment]
That's looking good!

**Comment by markhuot** [comment]
FYI, I'm still working on this. It's a fairly major rewrite, but I like where it's going. Here's an event listener that provides the schema for the first-party Dropdown field,

https://github.com/markhuot/craftql/blob/29aa9b3d0b7891fb2c620aca89fbbe5f9ab76d46/src/Listeners/GetSelectOneFieldSchema.php#L13-L23

Once this merges you should be able to avoid referencing GraphQL at all. I guess, you'd need to reference CraftQL, but you could do that in a single file that's only included if `class_exists(CraftQL::class)`

**Comment by markhuot** [comment]
This update is available in `1.0.0-beta.27`. I added a PR to SimpleMap to support the new syntax.

FYI, I'm not opposed to having the SimpleMap / CraftQL logic live in this repo. Whatever's easiest for you.
